parser/parser.y

```diff
@@ -3914,14 +3914,18 @@ CastType:
 		x := types.NewFieldType(mysql.TypeString)
 		x.Flen = $2.(int)
 		x.Charset = charset.CharsetBin
-		x.Collate = charset.CharsetBin
+		x.Collate = charset.CollationBin
 		$$ = x
 	}
 |	"CHAR" OptFieldLen OptBinary OptCharset
 	{
 		x := types.NewFieldType(mysql.TypeString)
 		x.Flen = $2.(int)
 		x.Charset = $4.(string)
+		if x.Charset == "" {
+			x.Charset = charset.CharsetUTF8
```

Comment:
`OptBinary OptCharset` in TiDB is the same as `opt_binary` in MySQL.
I add the binary flag when `OptBinary` is "BINARY" refer to the MySQL source code.